### PR TITLE
JIT: Implement greedy RPO-based block layout

### PR DESCRIPTION
**Pull request by amanasifkhalid**
Part of #93020. `Compiler::fgDoReversePostOrderLayout` reorders blocks based on a RPO of the flowgraph's successor edges. When reordering based on the RPO, we only reorder blocks within the same EH region to avoid breaking up their contiguousness. After establishing an RPO-based layout, we do another pass to move cold blocks to the ends of their regions in `fgMoveColdBlocks`.

The "greedy" part of this layout isn't all that greedy just yet. For now, we use edge likelihoods to make placement decisions only for `BBJ_COND` blocks' successors. I plan to extend this greediness to other multi-successor block kinds (`BBJ_SWITCH`, etc) in a follow-up so we can independently evaluate the value in doing so.

This new layout is disabled by default for now.

### CODE REVIEW

**Comment by amanasifkhalid** [comment]
cc @dotnet/jit-contrib, @AndyAyersMS PTAL. [Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=657574&view=ms.vss-build-web.run-extensions-tab) are big, though `libraries_tests` seems to be inflating the numbers. Improvements and regressions seem to be universally larger on x64 versus arm64 -- I assume this is because many short branches are becoming long on x64, and many long branches are becoming short, etc. Also, some of the arm64 example diffs suggest we ought to extend our emitter-level branch removal optimization beyond x64: I see some branches that we should be able to get rid of.

In my previous iterations of this PR, I wanted the block order to mirror the RPO as closely as possible, and repair try regions as necessary afterwards. I got this approach working when using funclets, but Windows x86's EH model complicated this by allowing handlers in the main method body too... I tried iterating on my initial implementation to handle the non-funclet EH case, but I was approaching two separate solutions, which didn't seem ideal -- plus all of these local fixups is something we're trying to get away from in the old layout approach. The diffs don't seem any more or less extreme by restricting reordering to within EH regions, and the code is quite a bit easier to follow.

**Comment by AndyAyersMS** [comment]
Seems likely that some significant part of the size improvement is from not moving cold blocks out of line. I think we may want this part to be implemented before we start running in the lab. It shouldn't be too hard to implement something like this: starting at the end of each EH region, move sufficiently cold blocks to the end, keeping them in relative RPO order.

I'm ok taking that as a follow up, if you think it makes sense. Though if it is indeed that simple it would not complicate this PR too much, and this PR is delightfully simple compared to the code it will replace.

You may recall in the example code I referred you as a model to there are some exceptions to this rule, and perhaps some thresholds to consider -- for instance sometimes we might not want to not move very small cold blocks that have hot pred and succ blocks, if we're fairly confident in the data, assuming we'd end up with denser code by keeping the branch short even if it means having a (well predicted) taken branch on the hot path.

**Comment by jakobbotsch** [comment]
Can you explain a bit further why you want the regular successor DFS? It saddens me a bit to see it return after the work I did last year to move everything to the faithful DFS. When I did that work I didn't find any justified uses of the old DFS.

For this case, why can't we run run the faithful DFS (say with your modifications to change the visit order) and then partition the RPO into multiple ones by EH region? I don't quite understand why we need to give up on all EH regions (or regions in the main body that are dominated by an EH region).

File: src/coreclr/jit/compiler.hpp
Comment (jakobbotsch) [reply src/coreclr/jit/compiler.hpp:729]: Ah, I misread. You do continue from the successor in the non-retless case.

**Comment by amanasifkhalid** [comment]
> Can you explain a bit further why you want the regular successor DFS?

My initial intent was to do a DFS of all successors without worrying about EH regions at all, and then evaluate how involved the needed EH fixups would be. Allowing the funclets to be broken up by the RPO complicated the post-ordering adjustments, as we'd have to take care not to make handler regions non-contiguous while fixing try regions, and vice versa. It seemed easier and not too consequential in terms of perf to just skip EH successors altogether, though the lack of funclets on Windows x86 made this problem unavoidable -- my attempts to appease win-x86 EH semantics were a distraction from the fact that I could just reorder within EH regions and not break anything up... I've pushed an implementation to try that out using a DFS of all successors.

> Seems likely that some significant part of the size improvement is from not moving cold blocks out of line (not splitting, per se, just separating within the method or region). I think we may want this part to be implemented before we start running in the lab.

I'll give that a try in my next push. I'll wait for the latest CI run to finish so we can see the updated diffs.

**Comment by amanasifkhalid** [comment]
@AndyAyersMS I tried renumbering the blocks after finalizing the layout locally, and the diffs are nontrivial, though they don't lean overwhelmingly in either direction. Here they are on win-x64:

<details>
<summary>Overall (<span style="color:green">-123,813</span> bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|PerfScore in Diffs
|---|--:|--:|--:|
|benchmarks.run.windows.x64.checked.mch|8,641,509|<span style="color:green">-95</span>|<span style="color:red">+0.03%</span>|
|benchmarks.run_pgo.windows.x64.checked.mch|33,740,443|<span style="color:green">-26,237</span>|<span style="color:green">-0.17%</span>|
|benchmarks.run_tiered.windows.x64.checked.mch|12,234,066|<span style="color:green">-2,303</span>|<span style="color:red">+0.07%</span>|
|coreclr_tests.run.windows.x64.checked.mch|402,587,216|<span style="color:red">+41,538</span>|<span style="color:red">+0.06%</span>|
|libraries.crossgen2.windows.x64.checked.mch|45,403,126|<span style="color:green">-353</span>|<span style="color:red">+0.06%</span>|
|libraries.pmi.windows.x64.checked.mch|62,248,989|<span style="color:green">-17,288</span>|<span style="color:red">+0.14%</span>|
|libraries_tests.run.windows.x64.Release.mch|280,210,097|<span style="color:green">-104,800</span>|<span style="color:green">-0.08%</span>|
|libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch|136,799,480|<span style="color:green">-11,192</span>|<span style="color:green">-0.03%</span>|
|realworld.run.windows.x64.checked.mch|13,554,143|<span style="color:green">-3,905</span>|<span style="color:red">+0.08%</span>|
|smoke_tests.nativeaot.windows.x64.checked.mch|5,016,576|<span style="color:red">+822</span>|<span style="color:red">+0.38%</span>|


</div></details>

<details>
<summary>FullOpts (<span style="color:green">-123,813</span> bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|PerfScore in Diffs
|---|--:|--:|--:|
|benchmarks.run.windows.x64.checked.mch|8,641,148|<span style="color:green">-95</span>|<span style="color:red">+0.03%</span>|
|benchmarks.run_pgo.windows.x64.checked.mch|20,302,793|<span style="color:green">-26,237</span>|<span style="color:green">-0.17%</span>|
|benchmarks.run_tiered.windows.x64.checked.mch|3,063,938|<span style="color:green">-2,303</span>|<span style="color:red">+0.07%</span>|
|coreclr_tests.run.windows.x64.checked.mch|122,068,398|<span style="color:red">+41,538</span>|<span style="color:red">+0.06%</span>|
|libraries.crossgen2.windows.x64.checked.mch|45,401,740|<span style="color:green">-353</span>|<span style="color:red">+0.06%</span>|
|libraries.pmi.windows.x64.checked.mch|62,135,488|<span style="color:green">-17,288</span>|<span style="color:red">+0.14%</span>|
|libraries_tests.run.windows.x64.Release.mch|105,590,714|<span style="color:green">-104,800</span>|<span style="color:green">-0.08%</span>|
|libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch|126,496,507|<span style="color:green">-11,192</span>|<span style="color:green">-0.03%</span>|
|realworld.run.windows.x64.checked.mch|13,148,501|<span style="color:green">-3,905</span>|<span style="color:red">+0.08%</span>|
|smoke_tests.nativeaot.windows.x64.checked.mch|5,015,585|<span style="color:red">+822</span>|<span style="color:red">+0.38%</span>|


</div></details>

<details>
<summary>Details</summary>
<div style="margin-left:1em">

#### Size improvements/regressions per collection

|Collection|Contexts with diffs|Improvements|Regressions|Same size|Improvements (bytes)|Regressions (bytes)|
|---|--:|--:|--:|--:|--:|--:|
|benchmarks.run.windows.x64.checked.mch|8,277|<span style="color:green">2,678</span>|<span style="color:red">2,051</span>|<span style="color:blue">3,548</span>|<span style="color:green">-19,743</span>|<span style="color:red">+19,648</span>|
|benchmarks.run_pgo.windows.x64.checked.mch|14,470|<span style="color:green">6,546</span>|<span style="color:red">3,062</span>|<span style="color:blue">4,862</span>|<span style="color:green">-48,223</span>|<span style="color:red">+21,986</span>|
|benchmarks.run_tiered.windows.x64.checked.mch|3,989|<span style="color:green">1,164</span>|<span style="color:red">725</span>|<span style="color:blue">2,100</span>|<span style="color:green">-7,177</span>|<span style="color:red">+4,874</span>|
|coreclr_tests.run.windows.x64.checked.mch|50,530|<span style="color:green">15,285</span>|<span style="color:red">25,649</span>|<span style="color:blue">9,596</span>|<span style="color:green">-101,366</span>|<span style="color:red">+142,904</span>|
|libraries.crossgen2.windows.x64.checked.mch|32,542|<span style="color:green">8,501</span>|<span style="color:red">8,938</span>|<span style="color:blue">15,103</span>|<span style="color:green">-55,666</span>|<span style="color:red">+55,313</span>|
|libraries.pmi.windows.x64.checked.mch|47,037|<span style="color:green">15,221</span>|<span style="color:red">13,968</span>|<span style="color:blue">17,848</span>|<span style="color:green">-115,545</span>|<span style="color:red">+98,257</span>|
|libraries_tests.run.windows.x64.Release.mch|64,836|<span style="color:green">23,951</span>|<span style="color:red">16,870</span>|<span style="color:blue">24,015</span>|<span style="color:green">-264,568</span>|<span style="color:red">+159,768</span>|
|libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch|66,421|<span style="color:green">19,152</span>|<span style="color:red">22,477</span>|<span style="color:blue">24,792</span>|<span style="color:green">-153,112</span>|<span style="color:red">+141,920</span>|
|realworld.run.windows.x64.checked.mch|9,577|<span style="color:green">3,105</span>|<span style="color:red">3,240</span>|<span style="color:blue">3,232</span>|<span style="color:green">-31,797</span>|<span style="color:red">+27,892</span>|
|smoke_tests.nativeaot.windows.x64.checked.mch|5,544|<span style="color:green">1,829</span>|<span style="color:red">1,816</span>|<span style="color:blue">1,899</span>|<span style="color:green">-12,544</span>|<span style="color:red">+13,366</span>|
||303,223|<span style="color:green">97,432</span>|<span style="color:red">98,796</span>|<span style="color:blue">106,995</span>|<span style="color:green">-809,741</span>|<span style="color:red">+685,928</span>|

---

#### PerfScore improvements/regressions per collection

|Collection|Contexts with diffs|Improvements|Regressions|Same PerfScore|Improvements (PerfScore)|Regressions (PerfScore)|PerfScore Overall in FullOpts|
|---|--:|--:|--:|--:|--:|--:|--:|
|benchmarks.run.windows.x64.checked.mch|8,277|<span style="color:green">1,563</span>|<span style="color:red">1,646</span>|<span style="color:blue">5,068</span>|<span style="color:green">-1.61%</span>|<span style="color:red">+1.69%</span>|<span style="color:red">+0.0082%</span>|
|benchmarks.run_pgo.windows.x64.checked.mch|14,470|<span style="color:green">3,354</span>|<span style="color:red">1,841</span>|<span style="color:blue">9,275</span>|<span style="color:green">-1.37%</span>|<span style="color:red">+1.20%</span>|<span style="color:green">-0.0518%</span>|
|benchmarks.run_tiered.windows.x64.checked.mch|3,989|<span style="color:green">577</span>|<span style="color:red">493</span>|<span style="color:blue">2,919</span>|<span style="color:green">-1.69%</span>|<span style="color:red">+2.56%</span>|<span style="color:red">+0.0163%</span>|
|coreclr_tests.run.windows.x64.checked.mch|50,530|<span style="color:green">5,662</span>|<span style="color:red">23,029</span>|<span style="color:blue">21,839</span>|<span style="color:green">-1.68%</span>|<span style="color:red">+0.54%</span>|<span style="color:red">+0.0121%</span>|
|libraries.crossgen2.windows.x64.checked.mch|32,542|<span style="color:green">6,305</span>|<span style="color:red">7,634</span>|<span style="color:blue">18,603</span>|<span style="color:green">-1.91%</span>|<span style="color:red">+1.85%</span>|<span style="color:red">+0.0066%</span>|
|libraries.pmi.windows.x64.checked.mch|47,037|<span style="color:green">11,118</span>|<span style="color:red">12,760</span>|<span style="color:blue">23,159</span>|<span style="color:green">-1.74%</span>|<span style="color:red">+2.09%</span>|<span style="color:red">+0.0216%</span>|
|libraries_tests.run.windows.x64.Release.mch|64,836|<span style="color:green">15,474</span>|<span style="color:red">10,501</span>|<span style="color:blue">38,861</span>|<span style="color:green">-1.34%</span>|<span style="color:red">+1.51%</span>|<span style="color:green">-0.0262%</span>|
|libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch|66,421|<span style="color:green">17,299</span>|<span style="color:red">15,269</span>|<span style="color:blue">33,853</span>|<span style="color:green">-1.21%</span>|<span style="color:red">+1.26%</span>|<span style="color:green">-0.0067%</span>|
|realworld.run.windows.x64.checked.mch|9,577|<span style="color:green">2,293</span>|<span style="color:red">2,749</span>|<span style="color:blue">4,535</span>|<span style="color:green">-1.37%</span>|<span style="color:red">+1.45%</span>|<span style="color:red">+0.0236%</span>|
|smoke_tests.nativeaot.windows.x64.checked.mch|5,544|<span style="color:green">1,217</span>|<span style="color:red">1,614</span>|<span style="color:blue">2,713</span>|<span style="color:green">-2.76%</span>|<span style="color:red">+3.48%</span>|<span style="color:red">+0.0664%</span>|

---

#### Context information

|Collection|Diffed contexts|MinOpts|FullOpts|Missed, base|Missed, diff|
|---|--:|--:|--:|--:|--:|
|benchmarks.run.windows.x64.checked.mch|27,811|4|27,807|2 (0.01%)|2 (0.01%)|
|benchmarks.run_pgo.windows.x64.checked.mch|94,917|48,136|46,781|14 (0.01%)|14 (0.01%)|
|benchmarks.run_tiered.windows.x64.checked.mch|53,445|37,214|16,231|9 (0.02%)|9 (0.02%)|
|coreclr_tests.run.windows.x64.checked.mch|583,206|347,435|235,771|50 (0.01%)|50 (0.01%)|
|libraries.crossgen2.windows.x64.checked.mch|277,538|15|277,523|31 (0.01%)|31 (0.01%)|
|libraries.pmi.windows.x64.checked.mch|313,896|6|313,890|156 (0.05%)|156 (0.05%)|
|libraries_tests.run.windows.x64.Release.mch|677,126|482,477|194,649|66 (0.01%)|66 (0.01%)|
|libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch|320,901|21,949|298,952|31 (0.01%)|31 (0.01%)|
|realworld.run.windows.x64.checked.mch|34,064|3|34,061|0 (0.00%)|0 (0.00%)|
|smoke_tests.nativeaot.windows.x64.checked.mch|31,814|12|31,802|18 (0.06%)|18 (0.06%)|
||2,414,718|937,251|1,477,467|377 (0.02%)|377 (0.02%)|


</div></details>

**Comment by amanasifkhalid** [comment]
Also, the [latest SPMI run](https://dev.azure.com/dnceng-public/public/_build/results?buildId=660261&view=ms.vss-build-web.run-extensions-tab) shows the ratio of size improvements to regressions decreasing, though we're missing some collections.

File: src/coreclr/jit/compiler.hpp
Comment (AndyAyersMS) [inline src/coreclr/jit/compiler.hpp:670]: Am I confused, or is this visiting the less likely successor first?

```suggestion
            else if (useProfile && (GetTrueEdge()->getLikelihood() > GetFalseEdge()->getLikelihood()))
```

File: src/coreclr/jit/fgopt.cpp
Comment (AndyAyersMS) [inline src/coreclr/jit/fgopt.cpp:3422]: You can just remove this whole `#ifdef DEBUG` bit, we will check the BB list at end of phase, which is imminent.

File: src/coreclr/jit/fgopt.cpp
Comment (AndyAyersMS) [inline src/coreclr/jit/fgopt.cpp:3429]: I assume it's unlikely that we won't move blocks here? The other variant tries to keep track, but I'm not sure it is worth the extra bookkeeping. All we are doing by sometimes returning false is speeding up the checked jit a tiny bit.

File: src/coreclr/jit/fgopt.cpp
Comment (AndyAyersMS) [inline src/coreclr/jit/fgopt.cpp:4556]: This is now a DFS of everything, right? Not just the main function?

File: src/coreclr/jit/fgopt.cpp
Comment (AndyAyersMS) [inline src/coreclr/jit/fgopt.cpp:4575]: Seems plausible that this map could be built up during the pass below, since the pair tail should follow the pair head in the RPO, so when we encounter the pair head neither block should have moved yet.

File: src/coreclr/jit/fgopt.cpp
Comment (AndyAyersMS) [inline src/coreclr/jit/fgopt.cpp:4731]: Am curious why this doesn't need to update the try region ends?

Since this is finding the last block in region, and then doing insert before, it seems like it might leave one hot block at the end?

File: src/coreclr/jit/fgopt.cpp
Comment (AndyAyersMS) [inline src/coreclr/jit/fgopt.cpp:4812]: Seems like we could handle these, by skipping the pair tail (like you do now) and then moving both if we want to move the pair head?

File: src/coreclr/jit/fgopt.cpp
Comment (AndyAyersMS) [inline src/coreclr/jit/fgopt.cpp:4762]: Maybe be a bit more specific? 
```suggestion
        // lastColdTryBlocks[X] will be null. We will check for this case in tryMovingBlocks below.)
```

File: src/coreclr/jit/fgopt.cpp
Comment (AndyAyersMS) [inline src/coreclr/jit/fgopt.cpp:4624]: Consider using `ArrayStack<struct>` for this data (where `struct` contains your 3 data items). It will usually be able to avoid heap allocation.

**Review by AndyAyersMS — COMMENTED** [review]
Overall I still really like the shape of this -- a bit concerned with the complexity of maintaining the EH invariants, but we can live with it.

Left you a few comments to consider.

File: src/coreclr/jit/compiler.hpp
Comment (amanasifkhalid) [reply src/coreclr/jit/compiler.hpp:670]: It looks unintuitive, but I think we need to flip the comparison so the DFS is in the order we want. Consider the following block list, pre-ordering:
```
BBnum BBid ref try hnd preds           weight   IBC [IL range]   [jump]                            [EH region]        [flags]
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
BB01 [0014]  1                             1        [???..???)-> BB02(1)                 (always)                     i keep internal
BB02 [0000]  1       BB01                  1    100 [000..00E)-> BB04(0),BB03(1)         ( cond )                     i IBC
BB03 [0010]  1       BB02                  0.50  50 [00D..00E)-> BB05(1)                 (always)                     i IBC nullcheck
BB04 [0011]  1       BB02                  0      0 [00D..00E)-> BB05(1)                 (always)                     i IBC rare
BB05 [0012]  2       BB03,BB04             1        [00D..017)                           (return)                     i hascall gcsafe
```

When processing `BB02`, if we visit `BB03` before `BB04`, then we end up with an RPO that looks something like `[<BB03's successors>, BB03, BB04, BB02, ...]`, so after layout, we get this:
```
BBnum BBid ref try hnd preds           weight   IBC [IL range]   [jump]                            [EH region]        [flags]
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
BB01 [0014]  1                             1        [???..???)-> BB02(1)                 (always)                     i keep internal
BB02 [0000]  1       BB01                  1    100 [000..00E)-> BB04(0),BB03(1)         ( cond )                     i IBC
BB04 [0011]  1       BB02                  0      0 [00D..00E)-> BB05(1)                 (always)                     i IBC rare
BB03 [0010]  1       BB02                  0.50  50 [00D..00E)-> BB05(1)                 (always)                     i IBC nullcheck
BB05 [0012]  2       BB03,BB04             1        [00D..017)                           (return)                     i hascall gcsafe
```

If we instead visit the less likely successor (`BB04`) first, we push the more likely successor `BB03` up to `BB02` in the RPO, and get this layout:
```
BBnum BBid ref try hnd preds           weight   IBC [IL range]   [jump]                            [EH region]        [flags]
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
BB01 [0014]  1                             1        [???..???)-> BB02(1)                 (always)                     i keep internal
BB02 [0000]  1       BB01                  1    100 [000..00E)-> BB04(0),BB03(1)         ( cond )                     i IBC
BB03 [0010]  1       BB02                  0.50  50 [00D..00E)-> BB05(1)                 (always)                     i IBC nullcheck
BB04 [0011]  1       BB02                  0      0 [00D..00E)-> BB05(1)                 (always)                     i IBC rare
BB05 [0012]  2       BB03,BB04             1        [00D..017)                           (return)                     i hascall gcsafe
```

File: src/coreclr/jit/fgopt.cpp
Comment (amanasifkhalid) [reply src/coreclr/jit/fgopt.cpp:3429]: That was my reasoning for hard-coding the return value.

File: src/coreclr/jit/fgopt.cpp
Comment (amanasifkhalid) [reply src/coreclr/jit/fgopt.cpp:4556]: Yes, updated.

File: src/coreclr/jit/fgopt.cpp
Comment (amanasifkhalid) [reply src/coreclr/jit/fgopt.cpp:4575]: Thanks for pointing this out; this seems to work locally. Just to make sure I don't have to add an edge case for this, we would never expect the RPO traversal to begin with a call-finally pair, right?

File: src/coreclr/jit/fgopt.cpp
Comment (amanasifkhalid) [reply src/coreclr/jit/fgopt.cpp:4731]: Yeah, I have the same reservations about this approach; Phoenix does something similar, though the last block in a region seems to have different semantic meaning there. I could just insert the cold blocks after the last block in the region to avoid potentially pushing a hot exit block to the end, though that might introduce back edges on the cold path -- is that ok?

Comment (amanasifkhalid) [reply src/coreclr/jit/fgopt.cpp:4723]: Got it, I've adjusted my approach to insert cold blocks after the end of the region, and then move the insertion point to the end as well, but only if it is cold.

File: src/coreclr/jit/fgopt.cpp
Comment (amanasifkhalid) [reply src/coreclr/jit/fgopt.cpp:4583]: This approach only hit one failure in a SPMI replay. I'm digging into this locally, but I adjusted my approach for now to (hopefully) be less expensive.

**Comment by amanasifkhalid** [comment]
@AndyAyersMS thank you for the review! [Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=662094&view=ms.vss-build-web.run-extensions-tab) are more conservative now that we're moving cold blocks, though still pretty big.

Since we have TP to spare, do you think we should renumber blocks post-layout in this PR? Or maybe in a follow-up?

**Comment by AndyAyersMS** [comment]
> Since we have TP to spare, do you think we should renumber blocks post-layout in this PR?

If we're going to do it eventually, we might as well do it now, one less churn event to sort through.

File: src/coreclr/jit/jitconfigvalues.h
Comment (AndyAyersMS) [inline src/coreclr/jit/jitconfigvalues.h:758]: Make sure to remember to undo this before merging.

File: src/coreclr/jit/fgopt.cpp
Comment (AndyAyersMS) [reply src/coreclr/jit/fgopt.cpp:4583]: Yeah I like this new approach of just walking the EH table.

In principle the call finally pair tail is dominated by the call finally pair head, however depending on how we do the DFS we may see "infeasible cross flow" .. say we have a finally with two callfinallies, it should not be possible to call via one callfinally and return to the other's tail, but may look feasible in the flow graph.

**Review by AndyAyersMS — APPROVED** [review]
Overall this looks good to me.

It's a bit sad that we can't treat the "not in try" region symmetrically with the try regions, but I think having a bit of duplicated logic is ok and we can sort this out later if we think it matters.

I would give @jakobbotsch a chance to sign off too.

File: src/coreclr/jit/jitconfigvalues.h
Comment (amanasifkhalid) [reply src/coreclr/jit/jitconfigvalues.h:758]: Got it. I'm gonna do one more CI run with block renumbering to compare the diffs. I'll mark as no-merge for now to remind myself.

**Comment by amanasifkhalid** [comment]
Here are the [diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=662720&view=ms.vss-build-web.run-extensions-tab) with block renumbering. Size improvements overall are slightly bigger, and TP is slightly improved, too.

File: src/coreclr/jit/compiler.hpp
Comment (amanasifkhalid) [reply src/coreclr/jit/compiler.hpp:670]: Agreed, I'll fix this in a follow-up PR. As for what the new abstraction should look like, would you prefer we move the `useProfile` check into `AllSuccessorEnumerator`, or even introduce a new enumerator like `ProfileGuidedSuccessorEnumerator`?

File: src/coreclr/jit/fgopt.cpp
Comment (jakobbotsch) [inline src/coreclr/jit/fgopt.cpp:4636]: Doesn't seem like we actually use `callFinallyPairs` as a map, so it could presumably just be an `ArrayStack` of pairs.

File: src/coreclr/jit/fgopt.cpp
Comment (jakobbotsch) [inline src/coreclr/jit/fgopt.cpp:4653]: Nit: I'd suggest indexing from the bottom instead, so that things end up stored in order instead of in reverse in this array.

**Review by jakobbotsch — APPROVED** [review]
LGTM. Some minor feedback that you can feel free to address as you see fit.

**Comment by amanasifkhalid** [comment]
@jakobbotsch @AndyAyersMS Thank you both for the reviews! [No diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=663252&view=ms.vss-build-web.run-extensions-tab) now that the new layout is disabled by default.

**Comment by amanasifkhalid** [comment]
Failures are #101721.

File: src/coreclr/jit/compiler.hpp
Comment (jakobbotsch) [reply src/coreclr/jit/compiler.hpp:670]: Perhaps introduce two instance initializer methods on `AllSuccessorEnumerator` and then pass some factory method to `fgRunDfs`? E.g. the normal use would be:
```
fgRunDfs([](SuccessorEnumerator* enumerator, BasicBlock* block) { enumerator->InitializeAllSuccs(block); }, ...);
```
and the block layout use could be
```
fgRunDfs([](SuccessorEnumerator* enumerator, BasicBlock* block) { enumerator->InitializeAllSuccsForBlockLayout(block); }, ...);
```

**Pull request merged**
